lineage/LineageTree.py

```diff
@@ -0,0 +1,285 @@
+        for cell in level:
+            parent_holder.add(self.output_lineage.index(cell.parent))
+        return parent_holder
+
+    def _find_leaves(self):
+        leaves = []
+        for cell in self.output_lineage:
+            if cell._isLeaf():
+                leaves.append(cell)
+        return leaves
+
+    def __repr__(self):
+        if self.prune_boolean:
+            s1 = "This tree is pruned. It is made of {} states.\n For each state in this tree: ".format(self.num_states)
+            s_list = []
+            for state in range(self.num_states):
+                s_list.append("\n \t There are {} cells of state {}".format(self.lineage_stats[state].num_pruned_lin_cells, state))
+            seperator = ', '
+            s2 = seperator.join(s_list)
+            s3 = ".\n This pruned tree has {} many cells in total".format(len(self.pruned_lin_list))
+            return s1 + s2 + s3
+        else:
+            s1 = "This tree is NOT pruned. It is made of {} states.\n For each state in this tree: ".format(self.num_states)
+            s_list = []
+            for state in range(self.num_states):
+                s_list.append("\n \t There are {} cells of state {}".format(self.lineage_stats[state].num_full_lin_cells, state))
+            seperator = ', '
+            s2 = seperator.join(s_list)
+            s3 = ".\n This UNpruned tree has {} many cells in total".format(len(self.full_lin_list))
+            return s1 + s2 + s3
+
+    def __str__(self):
+        if self.prune_boolean:
+            s1 = "This tree is pruned. It is made of {} states.\n For each state in this tree: ".format(self.num_states)
+            s_list = []
+            for state in range(self.num_states):
+                s_list.append("\n \t There are {} cells of state {}".format(self.lineage_stats[state].num_pruned_lin_cells, state))
+            seperator = ', '
+            s2 = seperator.join(s_list)
+            s3 = ".\n This pruned tree has {} cells in total".format(len(self.pruned_lin_list))
+            return s1 + s2 + s3
+        else:
+            s1 = "This tree is NOT pruned. It is made of {} states.\n For each state in this tree: ".format(self.num_states)
+            s_list = []
+            for state in range(self.num_states):
+                s_list.append("\n \t There are {} cells of state {}".format(self.lineage_stats[state].num_full_lin_cells, state))
+            seperator = ', '
+            s2 = seperator.join(s_list)
+            s3 = ".\n This UNpruned tree has {} cells in total".format(len(self.full_lin_list))
+            return s1 + s2 + s3
+        
+# tools for traversing trees
+
+def tree_recursion(cell, subtree):
+    """ a recursive function that traverses upwards from the leaf to the root. """
+    if cell._isLeaf():
+        return
+    subtree.append(cell.left)
+    subtree.append(cell.right)
+    tree_recursion(cell.left, subtree)
+    tree_recursion(cell.right, subtree)
+    return
```

Comment:
shouldnt this return subtree?

Reply:
nah

Reply:
it can, but since the subtree is already an input argument, it will be a mutable variable even after the function finished

Reply:
so its global can u point me to how this is so

Reply:
the thing that we append to (`subtree`) is also an argument of the function

see https://runestone.academy/runestone/books/published/thinkcspy/Lists/UsingListsasParameters.html

Reply:
So this is filling in the `subtree` variable in place?

Reply:
Yep. Most of tHMM works by modifying these large lists of `Nxk` arrays in place (like the `betas`, `gammas`, `NF`, etc). Those are just lists, where each list item is an `Nxk` (`N` is the number of cells in that lineage, `k` is the desired amount of states). These arrays get modified heavily over the analysis.

 This is just a smaller example of that where `subtree` is a growing list of cells that are the subtree of a cell.